pkg/cloudprovider/openstack.go

```diff
@@ -0,0 +1,848 @@
+// and as such does not test if multiple networks of the same address family are assigned to
+// the port, given that the calling method will already have done this.
+func (o *OpenStack) getIPsOnPort(p neutronports.Port) (int, int) {
+	ipv4UsedIPs := make(map[string]struct{})
+	ipv6UsedIPs := make(map[string]struct{})
+
+	for _, ip := range p.FixedIPs {
+		if utilnet.IsIPv4(net.ParseIP(ip.IPAddress)) {
+			ipv4UsedIPs[ip.IPAddress] = struct{}{}
+		} else {
+			ipv6UsedIPs[ip.IPAddress] = struct{}{}
+		}
+	}
+	for _, ip := range p.AllowedAddressPairs {
+		if utilnet.IsIPv4(net.ParseIP(ip.IPAddress)) {
+			ipv4UsedIPs[ip.IPAddress] = struct{}{}
+		} else {
+			ipv6UsedIPs[ip.IPAddress] = struct{}{}
+		}
+	}
+
+	return len(ipv4UsedIPs), len(ipv6UsedIPs)
+}
+
+// reserveNeutronIPAddress creates a new unattached neutron port with the given IP on
+// the given subnet. This will server as our IPAM as it is impossible to create 2 ports
+// with the same IP on the same subnet. The created port will be identified with a custom
+// json encoded description field.
+// NOTE: We are not using tags. According to the neutron API, it's possible to add a tag when creating
+// a port. But gophercloud does not allow us to do that and we must use a 2 step process (create port, then
+// add tag). Therefore, putting what we need into the Description field allows for atomic port creation and
+// for more flexibility thanks to the JSON encoded message.
+func (o *OpenStack) reserveNeutronIPAddress(s neutronsubnets.Subnet, ip net.IP, nodeName string) (*neutronports.Port, error) {
+	// JSON encode the EgressIP description first. This allows us to identify our
+	// own ports for retrieval and deletion later down the road.
+	var d []byte
+	description := EgressIPDescription{
+		Type:     egressIPTag,
+		NodeName: nodeName,
+	}
+	d, err := json.Marshal(description)
+	if err != nil {
+		return nil, fmt.Errorf("could not marshal object %v into JSON for the IP address reservation process", description)
+	}
+
+	// Now, create the port.
+	opts := neutronports.CreateOpts{
+		NetworkID: s.NetworkID,
+		FixedIPs: []neutronports.IP{
+			{
+				SubnetID:  s.ID,
+				IPAddress: ip.String(),
+			},
+		},
+		Description: string(d),
+		Name:        fmt.Sprintf("egressip-%s", ip.String()),
+	}
+	p, err := neutronports.Create(o.neutronClient, opts).Extract()
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+// releaseNeutronIPAddress deletes an unattached neutron port with the given IP on
+// the given subnet. It also looks at the description field and makes sure that the port matches.
+func (o *OpenStack) releaseNeutronIPAddress(port neutronports.Port, nodeName string) error {
+	if port.DeviceID != "" || port.DeviceOwner != "" {
+		return fmt.Errorf("port '%s' belongs to another device (%s) and/or device owner (%s), cannot delete it",
+			port.ID, port.DeviceID, port.DeviceOwner)
+	}
+
+	has, err := hasNodeName(port.Description, nodeName)
```

Comment:
Hm, using a description as a semantic field is not great. I would say `DeviceID` and `DeviceOwner` would be a perfect match to hold this info instead. These can be populated by the non-admin users as this worked on my cloud:

```
openstack port create --network db651749-4c98-41b0-897e-ebe547b66678 --device-owner "OpenShiftEgressIP" --device "ostest-xb42n-worker-0-4bhp8" test
```

We support having multiple clusters installed in a single OpenStack project, so you might want to include cluster ID into the `DeviceOwner` field. Please note that IIRC these are 255 VARCHAR fields, you might want to double-check and trim it.

Reply:
The neutron API itself does not list any limits: https://docs.openstack.org/api-ref/network/v2/index.html?expanded=update-port-detail
I'll have to check if the backing mysql uses a text field or a varchar

Reply:
`node.Status.NodeInfo.MachineID` contains the OpenStack Instance UUID. That should solve the issue for multiple clusters inside the same project and for the VARCHAR(255)

Reply:
I agree, using instance ID makes sense.